File: it-agents/skills/code-architect/references/code-examples/cpp/Flyweight.cpp
// Flyweight — share intrinsic state across many objects to save memory.
#include <iostream>
#include <map>
#include <memory>
#include <string>

// Intrinsic (shared) state.
class Glyph {
    char symbol;
    std::string font;
public:
    Glyph(char s, std::string f) : symbol(s), font(std::move(f)) {}
    // Extrinsic state (position) is passed in at use time.
    void draw(int x, int y) const {
        std::cout << "'" << symbol << "'/" << font << " at (" << x << "," << y << ")\n";
    }
};

// Flyweight factory caches and reuses shared instances.
class GlyphFactory {
    std::map<std::string, std::shared_ptr<Glyph>> pool;
public:
    std::shared_ptr<Glyph> get(char symbol, const std::string& font) {
        std::string key = std::string(1, symbol) + ":" + font;
        auto it = pool.find(key);
        if (it == pool.end())
            it = pool.emplace(key, std::make_shared<Glyph>(symbol, font)).first;
        return it->second;
    }
    size_t distinct() const { return pool.size(); }
};

int main() {
    GlyphFactory factory;
    std::string text = "aba";
    int x = 0;
    for (char c : text) {
        factory.get(c, "serif")->draw(x, 0);
        x += 10;
    }
    std::cout << "distinct flyweights = " << factory.distinct() << "\n";
    return 0;
}

File: it-agents/skills/code-architect/references/code-examples/rust/flyweight.rs
// Flyweight — share immutable intrinsic state across many objects to save
// memory; extrinsic state is passed in at use time.

use std::collections::HashMap;
use std::rc::Rc;

// Intrinsic, shareable state (the flyweight).
#[derive(Debug)]
struct Glyph {
    symbol: char,
    width: u32,
}

// Factory hands out shared flyweights, creating each only once.
struct GlyphFactory {
    pool: HashMap<char, Rc<Glyph>>,
}

impl GlyphFactory {
    fn new() -> Self {
        GlyphFactory { pool: HashMap::new() }
    }

    fn get(&mut self, symbol: char) -> Rc<Glyph> {
        self.pool
            .entry(symbol)
            .or_insert_with(|| {
                println!("creating new glyph for '{}'", symbol);
                Rc::new(Glyph { symbol, width: 8 })
            })
            .clone()
    }
}

fn main() {
    let mut factory = GlyphFactory::new();
    let text = "banana";

    // Extrinsic state (position) lives with each placement, glyphs are shared.
    let placed: Vec<(usize, Rc<Glyph>)> =
        text.chars().enumerate().map(|(pos, c)| (pos, factory.get(c))).collect();

    for (pos, glyph) in &placed {
        println!("pos {} -> '{}' (w={})", pos, glyph.symbol, glyph.width);
    }
    println!("unique glyphs allocated: {}", factory.pool.len());
}
